### DISCUSSION TITLE
TST recovery/backup session data

### DISCUSSION THREAD

**Discussion by davidmbesonen** [piroor/treestyletab#3003] [Q&A]
[TST 3.8.12, Firefox 88.0.1 (64-bit), Ubuntu Studio 20.04]

Does TST create recovery/backup data that could possibly be used to recover the state of a previous session of tabs?

The situation I'd like to recover from is one where Firefox crashed and after restarting Firefox, Firefox displayed all the pre-crash tabs minus any TST data (IOW, zero indentation).  I have refrained from using Firefox in the hope that there might exist some pre-crash TST data that I can retrieve and use to restore the pre-crash TST state.

**Comment by irvinm** [comment]
TST shouldn't lose it's indentation information even if Firefox crashes, but clearly you are seeing that behavior.  One way to solve this would be to use a session manager like "Tab Session Manager" which supports storing the TabOpenerID which basically means it supports storing the tab indentation information and therefore can restore all of your relationships if needed.

**Reply by davidmbesonen** [reply]
> One way to solve this would be to use a session manager like "Tab Session Manager"

I was thinking about doing this to backup a set of TST tab relationships. I'm glad to understand that this is likely to be an effective workaround.

It does make me wonder if this functionality might be built into TST? I would love it if TST was automatically creating a periodic rock-solid backup of it's current session.

**Reply by davidmbesonen** [reply]
> It does make me wonder if this functionality might be built into TST? I would love it if TST was automatically creating a periodic rock-solid backup of it's current session.

Another possibility is to simply have TST automatically bookmark all tabs on a periodic basis.

TST again failed to restore tab relationship data for me this morning.  Luckily, I used the "Bookmark this Tree..." option last night. So I was able to close the window with the lost tab relationship data, and then restore the tabs (and relationship data) from bookmarks.

I assume there is currently no built-in function (or TST addon) that would periodically backup all tabs to bookmarks?

**Comment by piroor** [comment]
Tree structure information of TST is stored to Firefox's session data via WebExtensions API, so all restored tabs should have restored tree information. But by some reasons TST can fail to load restored session information due to Firefox's restriction. Here is a possible workaround on such case:

1. Open one more new window. (Ctrl-N)
2. Close the restored window.
3. Reopen the window closed at the step 2. (Ctrl-Shift-N)

With these steps you possibly see restored tree. If you still don't get the result, sadly we have no way to get the lost tree information back.
